walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,73 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.lang.Math;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+        
+        Collection<String> attendees = request.getAttendees();
+        if (attendees.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        ArrayList<TimeRange> attendeeMeetingTimeRanges = 
+            getAttendeeMeetingTimeRanges(events, attendees);
+        Collections.sort(attendeeMeetingTimeRanges, TimeRange.ORDER_BY_START);
+
+        ArrayList<TimeRange> availableTimeRanges = new ArrayList<TimeRange>();
+        int start = TimeRange.START_OF_DAY;
+        int idx = 0;
+        while (idx < attendeeMeetingTimeRanges.size()) {
+            while ((idx < attendeeMeetingTimeRanges.size()) && 
+                (start >= attendeeMeetingTimeRanges.get(idx).start())) {
+                start = Math.max(start, attendeeMeetingTimeRanges.get(idx).end());
+                idx++;
+            }
+            if (idx >= attendeeMeetingTimeRanges.size()) break;
+
+            if (!attendeeMeetingTimeRanges.get(idx).contains(start)) {
```

Comment:
CMIIW, the conditions to reach this if-statement is

1. idx < size()
2. start < attendeeMeetingTimeRanges.get(idx).start()

So why do you need this if-statement?

Reply:
There is no need for if-statement here.